### ISSUE TITLE
AST of import declaration without attributes

### ISSUE THREAD

**Issue by sosukesuzuki** [estree/estree#314]
The following AST is defined for import attributes:

```ts
extend interface ImportDeclaration {
    attributes: [ ImportAttribute ];
}
```

This means `import foo from "mod";` and `import foo from "mod" with {};` have the same AST.

This isn't good for implementing code formatters. For example, Prettier's implementation reads the original text of the input to print attributes( https://github.com/prettier/prettier/blob/f5e339351e792cea0e8a6b111b336d5efa6441ea/src/language-js/print/module.js#L251-L272 ).

So I propose representing `attributes` as an empty array in `import foo from "mod" with {};` and as `null` in `import foo from "mod";`.

What do you think?

**Comment by bradzacher** [comment]
One point *against* doing this is that the import specifiers follow the same pattern - the array is always defined, regardless of the presence of the braces. So keeping it like this is standardised with that, at least, for better or worse.

Another point against this is that there are already implementations of the current spec. Things can change though and (AFAIK) acorn/eslint don't have support yet.

**Comment by sosukesuzuki** [comment]
> One point against doing this is that the import specifiers follow the same pattern - the array is always defined, regardless of the presence of the braces. So keeping it like this is standardised with that, at least, for better or worse.

Yes, Prettier actually determines whether to print `{}` for import specifiers in the same way ( https://github.com/prettier/prettier/blob/6db66c53d948d0a70dbc7abfa3f7d6f01f544242/src/language-js/print/module.js#L233-L249 ). If it's still possible to change the spec at this point, as someone implementing a code formatter, I would appreciate it if the import specifiers for `import {} from "mod";` and `import "mod";` could be distinguished as an empty array and `null`.

> Another point against this is that there are already implementations of the current spec. Things can change though and (AFAIK) acorn/eslint don't have support yet.

Since import attributes is still stage 3 syntax, I think we can still be able to change the estree spec.

**Comment by sosukesuzuki** [comment]
@JLHwung  What do you think?

**Comment by JLHwung** [comment]
I agree with @bradzacher, like he said the import specifiers follows the always-array pattern even if the empty braces `{}` are optional in the syntax. 

This rule is not limited to empty braces. For example ESTree does not differentiate `new A` with `new A()` as both of them provide an empty `argument` array. In this case, Prettier formats `new A` into `new A()`. Can prettier prefer `import foo from "foo"` over `import foo from "foo" with {}`?

**Comment by nzakas** [comment]
Agree with @bradzacher and @JLHwung here. ESTree is an abstract syntax tree, which necessarily means that not every token is represented, but rather, the *meaning* of the syntax is represented. I think the meaning is properly conveyed with the empty array.
